### PR TITLE
convert toir.c to D

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
Ready to go.

**Comment by WalterBright** [comment]
@rainers the 64 bit msvc build of dmd is failing - can you help with that?

**Comment by rainers** [comment]
> @rainers the 64 bit msvc build of dmd is failing - can you help with that?

Works here after after updating the VS projects. Where do you see the failure?

File: src/backend/cc.d [hunk 1]
```diff
@@ -1572,6 +1572,20 @@ extern __gshared EEcontext eecontext;
 //
 //extern Symbol *rtlsym[RTLSYM_MAX];
 
+enum
+{
+    RTLSYM_ALLOCMEMORY = 59,
+    RTLSYM_MAX = 147
+}
+
+// Workaround 2.066.x bug by resolving the TYMAX value before using it as dimension.
```
Comment (rainers) [inline src/backend/cc.d]: I guess you mean `RTLSYM_MAX` in the comment.

File: src/backend/el.c [hunk 2]
```diff
@@ -891,6 +891,8 @@ elem * el_longt(type *t,targ_llong val)
   return e;
 }
 
+extern "C"
+{
```
Comment (rainers) [inline src/backend/el.c]: This looks like a ugly workaround. Please add a comment why it is necessary.

**Comment by WalterBright** [comment]
> Works here after after updating the VS projects. Where do you see the failure?

 continuous-integration/travis-ci/pr — The Travis CI build failed

**Comment by rainers** [comment]
> continuous-integration/travis-ci/pr — The Travis CI build failed

That is not an msvc build, but OS/X. I don't have access to a Mac, but maybe the declarations for el_long actually don't match there?

**Comment by WalterBright** [comment]
@rainers Thanks for the tip, was able to reproduce it.

**Comment by rainers** [comment]
> @rainers Thanks for the tip, was able to reproduce it.

Good catch. I was rather hoping that you would have to make the prototype explicit as the use of extern(C) is pretty ugly here. Isn't it possible to specify the long argument for a C++ function on OSX/32? The tests suggest that it should work. Worst case, there would be two versioned declarations of `el_long`.

**Comment by WalterBright** [comment]
We have a problem in general with C++ name mangling for `long` because of its varying size. But this would be off-topic for this PR to do more than just get it to work. This PR should be solely about converting `toir.c` to `toir.d`, and it's ready to pull.

**Comment by WalterBright** [comment]
(BTW, the bug wasn't about `el_tolong`, but an alignment issue in `struct elem`.)

**Comment by andralex** [comment]
Kühl

**Pull request merged**